### PR TITLE
Add reverse iteration for btree db

### PR DESCRIPTION
**Pull request by i1i1**
This pr adds possibility to go back with btree iterator.

Reviewing by commits would make more sense, as the first two ones are minor refactoring. The rest of commits propagate reverse iterator from bottom to the top.

There are some minor problems with this code though. It repeats itself a lot, because I started of with just negating invariants of `next` function. Also, there is clearly not enough of tests for that, will add them soon.

### CODE REVIEW

File: src/btree/iter.rs [hunk 1]
```diff
@@ -155,17 +251,42 @@ impl<'a> BTreeIterator<'a> {
 		record_id: u64,
 		col: &BTreeTable,
 		log: &impl LogQuery,
-	) -> Result<Option<(Vec<u8>, Vec<u8>)>> {
+	) -> Result<Option<(Vec<u8>, Value)>> {
 		let BtreeIterBackend(tree, iter) = &mut self.iter;
-		if record_id != tree.record_id {
+		if record_id != tree.record_id || self.direction != IterDirection::Forward {
```
Comment (cheme) [inline src/btree/iter.rs:256]: Isn't change of direction handle at backend iterator level already (BTreeIterState)? (commit overlay should not need it)

File: src/btree/iter.rs [hunk 2]
```diff
@@ -202,7 +323,17 @@ impl BTreeIterState {
 		btree: &mut BTree,
 		col: &BTreeTable,
 		log: &impl LogQuery,
-	) -> Result<Option<(Vec<u8>, Vec<u8>)>> {
+		last_key: Option<impl AsRef<[u8]>>,
+		last_direction: IterDirection,
+	) -> Result<Option<(Vec<u8>, Value)>> {
+		if last_direction == IterDirection::Backward {
+			if let Some(last_key) = last_key {
+				self.seek(last_key.as_ref(), btree, col, log, SeekTo::After)?;
```
[the remark this answers is not in the document]
Comment (cheme) [reply src/btree/iter.rs:331]: Ok probably better sticking to `seek` in a first step.

File: src/btree/iter.rs [hunk 3]
```diff
@@ -237,8 +368,70 @@ impl BTreeIterState {
...
+
+		if self.next_separator && self.state.is_empty() {
+			return Ok(None)
+		}
+		if !self.next_separator {
+			if self.state.is_empty() {
+				let root = col.with_locked(|tables| {
+					BTree::fetch_root(btree.root_index.unwrap_or(NULL_ADDRESS), tables, log)
+				})?;
+				self.state.push((ORDER - 1, root));
```
[the remark this answers is not in the document]
Comment (cheme) [reply src/btree/iter.rs:398]: I think runing test with ORDER = 2 may show an issue but I did not try.

**Review by cheme — COMMENTED** [review]
Thanks for the PR, sorry for the lack of existing tests.

Generally testing here is a bit useless due to a btree ORDER being hardcoded to 8.
(running the iter test with ORDER = 2 makes more sense (code was written with 2 or 3 and then order was change for perf, but tests were not updated :( ).
parity-db-admin stress test can also be modified to check iteration value too and maybe do random seek (I remember randomly hitting seek in it but drop the code as it only work for small size db (reference values in memory)).
I guess middle ground would be a test with enough value to fill at least 3 level of nodes for current order 8 (eg 1024 insert followed by 512 removal) and then some random seek and iter.

File: src/btree/iter.rs [hunk 4]
```diff
@@ -237,8 +368,70 @@ impl BTreeIterState {
...
+		&mut self,
+		btree: &mut BTree,
+		col: &BTreeTable,
+		log: &impl LogQuery,
+		last_key: Option<impl AsRef<[u8]>>,
+		last_direction: IterDirection,
+	) -> Result<Option<(Vec<u8>, Value)>> {
+		if last_direction == IterDirection::Forward {
+			if let Some(last_key) = last_key {
+				self.seek(last_key.as_ref(), btree, col, log, SeekTo::Before)?;
```
[the remark this answers is not in the document]
Comment (i1i1) [reply src/btree/iter.rs:384]: Hmm, I thought that this way we can update backend's state (which assumed previously that we were doing forward iterations). How can I do that apart from seek?

File: src/btree/iter.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (i1i1) [reply src/btree/iter.rs:331]: No, it won't do. I believe that all the children are invalid. `Node::seek` depending on `from_end` flag will produce a totally different state:
https://github.com/paritytech/parity-db/pull/75/files#diff-69ea41d30427a984a750f3840d80a58558960cbc059024799e477f1dd0cd83c7R501

[the remark this answers is not in the document]
Comment (cheme) [reply src/btree/iter.rs:331]: Oh yes you're right, all parent in child need to be recalculated :+1:

File: src/btree/iter.rs [hunk 5]
```diff
@@ -26,6 +26,19 @@ use crate::{
...
 
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum IterDirection {
+	Backward,
+	Forward,
+}
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy)]
+pub enum SeekTo {
+	Before,
```
[the remark this answers is not in the document]
Comment (i1i1) [reply src/btree/iter.rs:37]: Yeah, it would be probably better to use `(IterDirection, SeekTo)` where `SeekTo` is either `At` or `After`! Thanks

File: src/btree/iter.rs (hunk 1 above)
Comment (i1i1) [reply src/btree/iter.rs:255]: Decided to update its state here and remove `BTreeIterState`'s direction update.

Otherwise, we would need to also pass `from_seek` to `BTreeIterState::next`, which would make more cubersome.

File: src/btree/iter.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (i1i1) [reply src/btree/iter.rs:379]: Hmm, Didn't get the issue here. I thought that accessible range is `0..ORDER` and `ORDER_CHILD` is out of range? :thinking:

[the remark this answers is not in the document]
Comment (cheme) [reply src/btree/iter.rs:379]: Yes the existing tests may not have enough value.
https://github.com/paritytech/parity-db/commit/2c0b88a99e36cf3f6de8b5084601e4fbce308331 will show issue even without reducing the order (pipeline for CommitOverlay is working).
In a ORDER 2 node you got child1|value1|child2|value2|child3 so we should start at index 2 (ORDER_CHILD - 1)

**Comment by cheme** [comment]
An idea for activity: bench a runtime calling `ext_hashing_twox_64_version_1` host function against a function runing the twox_64 in wasm.
(numbers should be very interesting)

**Comment by i1i1** [comment]
Sorry for taking that long. There were some issues after I implemented the test which @cheme suggested. Because of that, decided to change the iterator algorithm and make explicit difference between separator and child (that simplifies FSM even though it uses more space)

**Review by cheme — COMMENTED** [review]
Thanks, is it possible to include the tests from https://github.com/paritytech/parity-db/commit/2c0b88a99e36cf3f6de8b5084601e4fbce308331 to the PR (there is still some issue, fetch root is missing and when Order > 2 there is another issue I did not identify).

File: src/btree/iter.rs [hunk 6]
```diff
@@ -88,57 +102,137 @@ impl<'a> BTreeIterator<'a> {
...
+			},
+		}
+	}
+
+	pub fn prev(&mut self) -> Result<Option<(Vec<u8>, Vec<u8>)>> {
+		let col = self.col;
+
+		// Lock log over function call (no btree struct change).
+		let commit_overlay = self.commit_overlay.read();
+		self.from_seek |= self.direction == IterDirection::Forward;
```
[the remark this answers is not in the document]
Comment (i1i1) [reply src/btree/iter.rs:173]: Yeah, it allows updating `from_seek` in one place instead of many.

File: src/btree/iter.rs [hunk 7]
```diff
@@ -155,119 +249,188 @@ impl<'a> BTreeIterator<'a> {
...
+	) -> Result<Option<(Vec<u8>, Value)>> {
 		let BtreeIterBackend(tree, iter) = &mut self.iter;
-		if record_id != tree.record_id {
+		if record_id != tree.record_id || self.direction != IterDirection::Forward {
+			self.direction = IterDirection::Forward;
 			let new_tree = col.with_locked(|btree| BTree::open(btree, log, record_id))?;
 			*tree = new_tree;
 			if let Some(last_key) = self.last_key.as_ref() {
-				iter.seek(last_key.as_slice(), tree, col, log, true)?;
+				self.pending_next_backend = None;
```
[the remark this answers is not in the document]
Comment (i1i1) [reply src/btree/iter.rs:259]: Nope. `self.pending_next_backend` isn't part of `self.iter`

**Comment by cheme** [comment]
> Sorry for taking that long. There were some issues after I implemented the test which @cheme suggested. Because of that, decided to change the iterator algorithm and make explicit difference between separator and child (that simplifies FSM even though it uses more space)

No problem (usually I am the one that takes too long :).
Didn't read your explanation here, it is fine with me to store it at every state, even if I think every time we pop the state is the same (maybe a debug_assert could confirm it). Using an enum instead of the bool in next_separator looks way better.

File: src/btree/iter.rs [hunk 8]
```diff
@@ -155,119 +249,188 @@ impl<'a> BTreeIterator<'a> {
...
+pub enum NodeType {
+	Child,
+	Separator,
+}
+
+#[derive(Debug)]
 pub struct BTreeIterState {
-	state: Vec<(usize, Node)>,
-	next_separator: bool,
+	state: Vec<(usize, NodeType, Node)>,
```
[the remark this answers is not in the document]
Comment (i1i1) [reply src/btree/iter.rs:318]: It's kinda cleared\r this way. Anyway, we won't waste that much memory on it, but it would simplify algorithm a lot.

File: src/btree/iter.rs [hunk 9]
```diff
@@ -155,119 +249,188 @@ impl<'a> BTreeIterator<'a> {
...
+				*ty = NodeType::Child;
+
+				let key_query = TableKeyQuery::Fetch(None);
+				let r = col.get_at_value_index(key_query, address, log)?;
+				return Ok(r.map(|r| (key, r.1)))
 			}
 		}
+
 		self.state.pop();
-		self.next_separator = true;
```
[the remark this answers is not in the document]
Comment (i1i1) [reply src/btree/iter.rs:240]: I think no. It might be that it would be a separator after a `pop()`.

File: src/btree/node.rs [hunk 10]
```diff
@@ -27,13 +30,12 @@ use crate::{
...
 impl Node {
-	fn last_separator_index(&self) -> Option<usize> {
-		let i = self.number_separator();
-		if i == 0 {
-			None
-		} else {
-			Some(i - 1)
-		}
+	pub(crate) fn last_child_index(&self) -> Option<usize> {
+		self.children.iter().rposition(|child| child.entry_index.is_some())
```
[the remark this answers is not in the document]
Comment (i1i1) [reply src/btree/node.rs:34]: Yeah, we want to find last entry which `is_some`. I thought about using binary search here, but it's kinda out of pr's scope anyway.

File: src/btree/iter.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (cheme) [reply src/btree/iter.rs:318]: yes agree it simplify things. 
(I double check, everytime push is called last state is indeed updated with `*ty = NodeType::Separator;` so updating on pop to `Separator` would certainly work)

File: src/btree/iter.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (cheme) [reply src/btree/iter.rs:240]: Actually it works with == NodeType::Separator (did not read all the code when writing the comment, I was expected that type was not updated when pushing item on stack

**Comment by cheme** [comment]
Another point that would need checking (I will be a bit offline nextweek so s I put the info rn even if a bit incomplete):
https://github.com/i1i1/parity-db/compare/reverse-iteration...cheme:parity-db:75-3?expand=1
is failing because when we restart backend iteration, the commit index is different, so the iterator does seek again against it's latest returned value. The issue here is that `seek_from` is set to `true` and it will return the last value again (it should be false so we know seek is note from an explicit seek call).
`seek_from` is true because during init the iterator change direction, which put its state to seek_from true. I guess something similar could be achieve in forward iteration by switching direction.
Generally switching direction of iterator is nice, but the implementation involves a seek, an idea to simplify could be to have a reverse iterator and a normal iterator only, so we do not need to manage change of iterator direction (user could still do it by seeking and skipping first value if equal to last value as I do with the btreemap in test).

File: src/db.rs [hunk 11]
```diff
@@ -1897,5 +1999,13 @@ mod tests {
 			let iter_next = iter.next().unwrap();
 			assert_eq!(state_next, iter_next.as_ref().map(|(k, v)| (k, v)));
 		}
+
+		let mut iter_state_rev = end_state.iter().rev();
+		let mut iter = db.iter(col_nb).unwrap();
+		iter.seek(&[0xff; 4]).unwrap();
```
[the remark this answers is not in the document]
Comment (i1i1) [reply src/db.rs:2005]: I though about adding `seek_to_first` and `seek_to_last` methods. First one can be achieved via `seek(&[])`, while second one requires more logic.

But overall I don't know if `iter` should start with both first and last item depending on first call of `prev` or `next`

[the remark this answers is not in the document]
Comment (cheme) [reply src/db.rs:2005]: Yes that's a bit questionable indeed. Usually if first call is prev I want to start from end and when first call is next we want to start from start, but that is not really the semantic of prev.
In this case would make sense to implement DoubleEndedIterator (with result<option as item) and `seek_to_first`, `seek_to_last`  as you propose.

**Comment by cheme** [comment]
One last small idea, the state could switch from :
```
	state: Vec<(usize, NodeType, Node)>,
```
to 
```
	state: Vec<(Option<usize>, NodeDirection, Node)>,
```
that way iterator could change direction without calling `seek`. But this also requires to change stored index to be:  last child index used for non last_mut (already is) or can also be last separator index returned (not already is).
Would make more lookalike implementation between forward and backward, actually could have a single function doing it by having a state compatible with both.

About `from_seek` , it should not be use for backend (backend see change of direction or change of commit id and know it should seek and always will seek after the last value).
It's use is related to commit overlay where we always take item after last_key returned, except if a seek was called, then it is correct to return the same key (last_key being initialized to the seek key). For backend we got state right when calling seek.
Generally the way I use `from_seek` is not good code (for instance after any call to btree_next it should be set to false instead of the multiple assignment currently in place).
Using this `from_seek` mechanism was because I did not want to write `previous_key` function and just store `last_key = Some(previous_key(seek_key))`, but I think it would have been a way better idea.

**Comment by cheme** [comment]
@i1i1 , sorry for all the feedback, lot of it might sounds a bit gibberish, it is me lacking retrospect end of week.

I think most of it could be TODO "later" (supporting change of direction without seek, merging prev and next backend logic, using doublenedediter).

Looks like things work properly with the PR (I really like the new test, could maybe use a seedable rng to have something reproducible on error), except the corner case which is a bit undefined situation anyway.

So last things really needed would be:
- fix the issue with `from_seek`.
- implement seek_end
- maybe move position_rev code to seek_prev (so in general position should be use).

For the `from_seek` part we could start with this (only make the code for master):
https://github.com/paritytech/parity-db/commit/91c67cc362ee61e1680809968b4293133bb10c24
makes my code way less bad that it was.


Ps: this is a draft impl from a test branch for seek_end (backend part), but it can certainly be written a bit better:
```

	fn reset_to_root(&mut self,
		btree: &mut BTree,
		col: &BTreeTable,
		log: &impl LogQuery,
	) -> Result<()> {
		if self.state.len() > 0 {
			self.state.truncate(1);
		} else {
			let root = col.with_locked(|tables| {
				BTree::fetch_root(btree.root_index.unwrap_or(NULL_ADDRESS), tables, log)
			})?;
			self.state.push((0, NodeType::Child, root));
		}
			Ok(())
	}

	pub fn seek_start(&mut self,
		btree: &mut BTree,
		col: &BTreeTable,
		log: &impl LogQuery,
	) -> Result<()> {
		self.reset_to_root(btree, col, log)?;
		self.state.last_mut().map(|(index, node_type, _)| {
			*index = 0;
			*node_type = NodeType::Child;
		});
		Ok(())
	}

	pub fn seek_end(&mut self, 
		btree: &mut BTree,
		col: &BTreeTable,
		log: &impl LogQuery,
	) -> Result<()> {
		self.reset_to_root(btree, col, log)?;
		self.state.last_mut().map(|(index, node_type, node)| {
			*index = node.last_child_index().unwrap_or(node.last_separator_index().unwrap_or(0));
			*node_type = NodeType::Child;
		});
		Ok(())
	}
```

**Comment by i1i1** [comment]
Sorry for the delay again. Just fixed the issue related to the [test which you pointed](https://github.com/i1i1/parity-db/compare/reverse-iteration...cheme:parity-db:75-3?expand=1) and added seek to the start and to the end. I can work on some of the ideas which you pointed, but not as the part of this pr 😅

**Review by cheme — COMMENTED** [review]
LGTM, added just a few question to be sure.
I may switch the from_seek to what I propose here https://github.com/paritytech/parity-db/commit/91c67cc362ee61e1680809968b4293133bb10c24 in the future (really not sure when), but I don't think it changes the current api. So we should be able to publish this as new minor version before.

File: src/btree/node.rs [hunk 12]
```diff
@@ -561,6 +561,32 @@ impl Node {
...
+		} else {
+			return Ok(())
+		};
+
+		if depth != 0 {
+			if let Some(child) = from.fetch_child(i, values, log)? {
+				if i > 0 {
+					stack.push((i - 1, NodeType::Separator, from));
+				}
+				return Self::seek_to_last(child, values, log, depth - 1, stack)
```
[the remark this answers is not in the document]
Comment (i1i1) [reply src/btree/node.rs:599]: Hmm, I believe it will also work without recursing but it will be lazy. I can change that if you insist, it will make code simpler I think :man_shrugging:

File: src/db.rs [hunk 13]
```diff
@@ -2016,7 +2016,6 @@ mod tests {
 
 		let mut iter_state_rev = end_state.iter().rev();
 		let mut iter = db.iter(col_nb).unwrap();
-		iter.seek(&[0xff; 4]).unwrap();
```
[the remark this answers is not in the document]
Comment (i1i1) [reply src/db.rs]: Yeah, but if you ask me, I would prefer having explicit `seek_to_last` call

[the remark this answers is not in the document]
Comment (cheme) [reply src/db.rs]: Yes please switch back to it if you prefer.

**Comment by cheme** [comment]
small changes needed:
```diff --cc src/db.rs
index 1e436c0,86ae052..0000000
--- a/src/db.rs
+++ b/src/db.rs
diff --git a/src/btree/iter.rs b/src/btree/iter.rs
index 7d707f9..741ec7d 100644
--- a/src/btree/iter.rs
+++ b/src/btree/iter.rs
@@ -408,7 +408,6 @@ impl BTreeIterState {
                        return Ok(None)
                }
 
-               dbg!(&self.state);
                while let Some((ix, ty @ NodeType::Child, node)) = self.state.last_mut() {
                        *ty = NodeType::Separator;
                        match col.with_locked(|btree| node.fetch_child(*ix, btree, log))? {
diff --git a/src/btree/node.rs b/src/btree/node.rs
index 7a0d4dd..9ad844a 100644
--- a/src/btree/node.rs
+++ b/src/btree/node.rs
@@ -585,7 +585,9 @@ impl Node {
                _depth: u32,
                stack: &mut Vec<(usize, NodeType, Self)>,
        ) -> Result<()> {
-               if let Some(i) = from.last_child_index().or_else(|| from.last_separator_index()) {
+               if let Some(i) = from.last_child_index() {
+                       stack.push((i, NodeType::Child, from))
+               } else if let Some(i) = from.last_separator_index() {
                        stack.push((i, NodeType::Separator, from))
                }
                Ok(())
```

**Comment by i1i1** [comment]
Should I just always fetch the last separator instead? It should do the same job, I guess.

**Comment by cheme** [comment]
it fails in some case (I did run it on a modified version of the stress test in https://github.com/cheme/parity-db/commits/75-tests (this is the fixed version) with `cargo run --release -p parity-db-admin -- stress -d ./t --commits 10000 --ordered --iter 2` if you want to take a look).

**Comment by cheme** [comment]
Wait I don't see it failing anymore, will check a bit more.

**Review by cheme — APPROVED** [review]
LGTMerge cc @arkpar

File: src/db.rs [hunk 14]
```diff
@@ -1448,23 +1469,32 @@ mod tests {
...
 
 		assert_eq!(db.get(col_nb, &key1).unwrap(), Some(b"value1".to_vec()));
-		iter.seek(&[]).unwrap();
+		iter.seek_to_first().unwrap();
 		assert_eq!(iter.next().unwrap(), Some((key1.clone(), b"value1".to_vec())));
 		assert_eq!(iter.next().unwrap(), None);
 
+		iter.seek_to_first().unwrap();
+		assert_eq!(iter.next().unwrap(), Some((key1.clone(), b"value1".to_vec())));
+		assert_eq!(iter.prev().unwrap(), Some((key1.clone(), b"value1".to_vec())));
```
Comment (cheme) [inline src/db.rs:1491]: @i1i1 question, shouldn't it be rather None?

Comment (i1i1) [reply src/db.rs:1491]: I believed that this is how it should behave, as after the call of `next` cursor was moved

Comment (cheme) [reply src/db.rs:1491]: I may change it to absolute cursor in the future (not sure when), then it would return None instead. Is it fine with you if I switch to this behavior in a future pr ?

Comment (i1i1) [reply src/db.rs:1491]: It would work either way for me before. The main thing will be to change semantics before the release :)
